### PR TITLE
feat: DODO V2, Linkswap

### PR DESCRIPTION
**Pull request by dekz**
## Description

Adds support for Dodo V2

```
SELL USDT->DODO 871.54 -> 130.39 ($871.54) after 0.1s @ dev
	✔ PASS 871.54 -> 130.39 ($832.97) ($52.41)
	DODO_V2: 100%
	gas: 235282
SELL DODO->USDT 61.23 -> 396.27 ($415.73) after 0.2s @ dev
	✔ PASS 61.23 -> 396.27 ($343.96) ($52.31)
	DODO_V2: 100%
	gas: 234854
SELL DODO->USDT 38.76 -> 251.62 ($263.18) after 0.8s @ dev
	✔ PASS 38.76 -> 251.62 ($199.30) ($52.31)
	DODO_V2: 100%
	gas: 234870
SELL DODO->USDT 142.55 -> 912.63 ($967.92) after 0.1s @ dev
	✔ PASS 142.55 -> 912.63 ($860.31) ($52.31)
	DODO_V2: 100%
	gas: 234870
```

```
BUY USDT->DODO 779.69 -> 116.87 ($804.07) after 0.4s @ dev
	✔ PASS 779.69 -> 116.87 ($851.69) ($72.00)
	DODO_V2: 100%
	gas: 267960
BUY DODO->USDT 32.88 -> 213.64 ($213.64) after 0.0s @ dev
	✔ PASS 32.88 -> 213.64 ($289.70) ($73.33)
	DODO_V2: 100%
	gas: 263372
```

## Testing instructions

Simbot config

```
yarn start-ab --url "dev=http://localhost:3001/swap/v1/quote?includedSources=DODO_V2" --output ab.json --jobs 1 --sells -p "DODO/USDT"
```

```json
{
    "erc20Proxy": "0x95e6f48254609a6ee006f7d493c8e5fb97094cef",
    "exchangeProxy": "0xdef1c0ded9bec7f1a1670819833240f027b25eff",
    "exchangeProxyOwnerStorageSlot": "0x0000000000000000000000000000000300000000000000000000000000000000",
    "taker": "0xd00d00caca000000000000000000000000001337",
    "gst": "0x0000000000b3F879cb30FE243b4Dfee438691c04",
    "transformers": {
        "deployer": "0x39dce47a67ad34344eab877eae3ef1fa2a1d50bb",
        "overridesByNonce": {
            "12": {
                "artifactPath": "../0x-monorepo/contracts/zero-ex/test/generated-artifacts/FillQuoteTransformer.json",
                "constructorArgs": [
                    "0xe3a207e4225d459095491ea75d30b31968dff887",
                    "0xdef1c0ded9bec7f1a1670819833240f027b25eff"
                ]
            }
        }
    },
    "overrides": {},
    "newFeatures": {
        "deployer": "0xfefefefefefefefefefefefefefefefefefefefe",
        "contracts": []
    },
    "deployments": {
        "deployer": "0xdededededededededededededededededededede",
        "initialNonce": 0,
        "contracts": [
            {
                "artifactPath": "../0x-monorepo/contracts/zero-ex/test/generated-artifacts/BridgeAdapter.json",
                "constructorArgs": [
                    "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2"
                ],
                "value": 0
            }
        ]
    }
}
```

## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: packages/asset-swapper/contracts/src/DODOV2Sampler.sol [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+            (address, address, bool)
+        );
+
+        // We will get called to sell both the taker token and also to sell the maker token
+        // since we use approximate buy for sell and buy functions
+        if (sellBase) {
+            try
+                IDODOV2Pool(pool).querySellBase
+                    { gas: DODO_V2_CALL_GAS }
+                    (address(0), sellAmount)
```
Comment (kimpers) [inline packages/asset-swapper/contracts/src/DODOV2Sampler.sol:159]: Silly question but what sorcery is this? Looks like a fn call without parentheses

Comment (dekz) [reply packages/asset-swapper/contracts/src/DODOV2Sampler.sol:159]: it's the special solidity syntax to set certain EVM arguments in how the function should be called. Things like gas (give is a gas stipend) and value (how much ETH (wei) to send).

It used to be more of a builder type pattern, until this syntax.

```
IDODOV2Pool(pool).querySellBase.gas(DODO_V2_CALL_GAS).value(MONEY)(address(0), sellAmount)
```

File: packages/asset-swapper/contracts/src/DODOV2Sampler.sol [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    {
+        _assertValidPair(makerToken, takerToken);
+
+        (pool, sellBase) = _getNextDODOV2Pool(registry, offset, takerToken, makerToken);
+        if (pool == address(0)) {
+            return (sellBase, pool, makerTokenAmounts);
+        }
+
+        uint256 numSamples = takerTokenAmounts.length;
+        makerTokenAmounts = new uint256[](numSamples);
```
Comment (dorothy-zbornak) [inline packages/asset-swapper/contracts/src/DODOV2Sampler.sol:84]: worth moving these up before the pool check so we don't return an empty array?

File: packages/protocol-utils/src/transformer_utils.ts [hunk 3]
```diff
@@ -127,6 +127,7 @@ export enum BridgeSource {
     Swerve,
     Uniswap,
     UniswapV2,
+    DodoV2,
```
Comment (dorothy-zbornak) [inline packages/protocol-utils/src/transformer_utils.ts:130]: Need changelog

**Review by dorothy-zbornak — APPROVED** [review]

**Pull request merged**